blockchain/blockchain.go

```diff
@@ -976,12 +981,30 @@ func (bc *BlockChain) GetLogsByHash(hash common.Hash) [][]*types.Log {
 	return logs
 }
 
-// TrieNode retrieves a blob of data associated with a trie node (or code hash)
+// TrieNode retrieves a blob of data associated with a trie node
 // either from ephemeral in-memory cache, or from persistent storage.
 func (bc *BlockChain) TrieNode(hash common.Hash) ([]byte, error) {
 	return bc.stateCache.TrieDB().Node(hash)
 }
 
+// ContractCode retrieves a blob of data associated with a contract hash
+// either from ephemeral in-memory cache, or from persistent storage.
+func (bc *BlockChain) ContractCode(hash common.Hash) ([]byte, error) {
+	return bc.stateCache.ContractCode(hash)
+}
+
+// ContractCodeWithPrefix retrieves a blob of data associated with a contract
+// hash either from ephemeral in-memory cache, or from persistent storage.
+//
+// If the code doesn't exist in the in-memory cache, check the storage with
+// new code scheme.
+func (bc *BlockChain) ContractCodeWithPrefix(hash common.Hash) ([]byte, error) {
+	type codeReader interface {
+		ContractCodeWithPrefix(codeHash common.Hash) ([]byte, error)
+	}
+	return bc.stateCache.(codeReader).ContractCodeWithPrefix(hash)
```

Comment:
It seems little weird. bc.stateCache is already an interface, do we need to cast this to another interface?
anyway, it seems bc.stateCache has always the interface `ContractCodeWithPrefix`. Otherwise, the casting will fail.

I'm not sure this porting should be changed, but I just wanted to discuss why we made an interface specifically on this.

Reply:
In my opinion, the `stateCache` should have `ContractCode` method only in the long term. I thought that this interface `codeReader` is defined here temporary, but it will be deleted at the end (not yet deleted though). Does this make sense?